src/Google.Api.Gax/ApiCallRetryExtensions.cs

```diff
@@ -0,0 +1,103 @@
+﻿using Google.Protobuf;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax
+{
+    public static class ApiCallRetryExtensions
+    {
+        public static ApiCall<TRequest, TResponse> WithRetry<TRequest, TResponse>(
+            this ApiCall<TRequest, TResponse> originalCall,
+            RetrySettings retrySettings,
+            IClock clock,
+            IScheduler scheduler)
+            where TRequest : class, IMessage<TRequest>
+            where TResponse : class, IMessage<TResponse>
+        {
+            if (retrySettings == null)
+            {
+                // If retry isn't required, simply return the original call.
+                return originalCall;
+            }
+
+            GaxPreconditions.CheckNotNull(originalCall, nameof(originalCall));
+            GaxPreconditions.CheckNotNull(retrySettings, nameof(retrySettings));
+            retrySettings.Validate(nameof(retrySettings));
+            GaxPreconditions.CheckNotNull(clock, nameof(clock));
+            scheduler = scheduler ?? SystemScheduler.Instance;
+
+            ApiCall<TRequest, TResponse>.AsyncCall asyncCall = async (request, callSettings) =>
```

Comment:
I'd possibly add a comment explaining the code duplication. (It's probably still the right decision, but worth explaining it.)